From 41fe52b60c5a971e86d608794161e65a4c5fc2df Mon Sep 17 00:00:00 2001
From: Kishan Kunduru <kkunduru@google.com>
Date: Wed, 15 Jun 2022 21:31:49 +0000
Subject: [PATCH 1/2] curl: Add separate speedtest mode flags

Separate out downlink and uplink speedtest flags. By default, speedtest
mode is off and only set when CURLOPT_DL_SPEEDTEST_MODE or
CURLOPT_UL_SPEEDTEST_MODE is enabled.

BUG: 236173401
TEST: Compile and verify default functionality is not changed and
speedtest mode are set properly.

Change-Id: I57a04aadeb7d48103f1dd3688792e066cbe6150e
---
 include/curl/curl.h | 7 +++++--
 lib/connect.c       | 2 +-
 lib/easyoptions.c   | 3 ++-
 lib/setopt.c        | 9 +++++++--
 lib/transfer.c      | 4 ++--
 lib/urldata.h       | 6 ++++--
 6 files changed, 21 insertions(+), 10 deletions(-)

diff --git a/include/curl/curl.h b/include/curl/curl.h
index 139617cd0..6df5c9eec 100644
--- a/include/curl/curl.h
+++ b/include/curl/curl.h
@@ -2105,8 +2105,11 @@ typedef enum {
      this option is used only if PROXY_SSL_VERIFYPEER is true */
   CURLOPT(CURLOPT_PROXY_CAINFO_BLOB, CURLOPTTYPE_BLOB, 310),
 
-  /* SpeedTest Mode, i.e discard read data and set performance flags */
-  CURLOPT(CURLOPT_SPEEDTEST_MODE, CURLOPTTYPE_LONG, 311),
+  /* DL SpeedTest Mode, i.e discard read data and set performance flags */
+  CURLOPT(CURLOPT_DL_SPEEDTEST_MODE, CURLOPTTYPE_LONG, 311),
+
+  /* UL SpeedTest Mode, i.e use zerocopy and MSG_DONTWAIT flags */
+  CURLOPT(CURLOPT_UL_SPEEDTEST_MODE, CURLOPTTYPE_LONG, 312),
 
   CURLOPT_LASTENTRY /* the last unused */
 } CURLoption;
diff --git a/lib/connect.c b/lib/connect.c
index 002e4c13c..00aca5b45 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -1344,7 +1344,7 @@ static CURLcode singleipconnect(struct Curl_easy *data,
 
   if(!result) {
     *sockp = sockfd;
-    if(data->set.speedtest_mode) {
+    if(data->set.dl_speedtest_mode) {
       int rcvlowat = data->set.buffer_size;
       if (setsockopt(sockfd, SOL_SOCKET, SO_RCVLOWAT,
                      &rcvlowat, sizeof(rcvlowat)) == -1) {
diff --git a/lib/easyoptions.c b/lib/easyoptions.c
index 590ec78b9..6bd2ceb68 100644
--- a/lib/easyoptions.c
+++ b/lib/easyoptions.c
@@ -72,6 +72,7 @@ struct curl_easyoption Curl_easyopts[] = {
   {"DIRLISTONLY", CURLOPT_DIRLISTONLY, CURLOT_LONG, 0},
   {"DISALLOW_USERNAME_IN_URL", CURLOPT_DISALLOW_USERNAME_IN_URL,
    CURLOT_LONG, 0},
+  {"DL_SPEEDTEST_MODE", CURLOPT_DL_SPEEDTEST_MODE, CURLOT_LONG, 0},
   {"DNS_CACHE_TIMEOUT", CURLOPT_DNS_CACHE_TIMEOUT, CURLOT_LONG, 0},
   {"DNS_INTERFACE", CURLOPT_DNS_INTERFACE, CURLOT_STRING, 0},
   {"DNS_LOCAL_IP4", CURLOPT_DNS_LOCAL_IP4, CURLOT_STRING, 0},
@@ -177,7 +178,6 @@ struct curl_easyoption Curl_easyopts[] = {
   {"NOPROGRESS", CURLOPT_NOPROGRESS, CURLOT_LONG, 0},
   {"NOPROXY", CURLOPT_NOPROXY, CURLOT_STRING, 0},
   {"NOSIGNAL", CURLOPT_NOSIGNAL, CURLOT_LONG, 0},
-  {"SPEEDTEST_MODE", CURLOPT_SPEEDTEST_MODE, CURLOT_LONG, 0},
   {"OPENSOCKETDATA", CURLOPT_OPENSOCKETDATA, CURLOT_CBPTR, 0},
   {"OPENSOCKETFUNCTION", CURLOPT_OPENSOCKETFUNCTION, CURLOT_FUNCTION, 0},
   {"PASSWORD", CURLOPT_PASSWORD, CURLOT_STRING, 0},
@@ -327,6 +327,7 @@ struct curl_easyoption Curl_easyopts[] = {
   {"TRAILERFUNCTION", CURLOPT_TRAILERFUNCTION, CURLOT_FUNCTION, 0},
   {"TRANSFERTEXT", CURLOPT_TRANSFERTEXT, CURLOT_LONG, 0},
   {"TRANSFER_ENCODING", CURLOPT_TRANSFER_ENCODING, CURLOT_LONG, 0},
+  {"UL_SPEEDTEST_MODE", CURLOPT_UL_SPEEDTEST_MODE, CURLOT_LONG, 0},
   {"UNIX_SOCKET_PATH", CURLOPT_UNIX_SOCKET_PATH, CURLOT_STRING, 0},
   {"UNRESTRICTED_AUTH", CURLOPT_UNRESTRICTED_AUTH, CURLOT_LONG, 0},
   {"UPKEEP_INTERVAL_MS", CURLOPT_UPKEEP_INTERVAL_MS, CURLOT_LONG, 0},
diff --git a/lib/setopt.c b/lib/setopt.c
index 8adf96d45..eed430d64 100644
--- a/lib/setopt.c
+++ b/lib/setopt.c
@@ -2210,10 +2210,15 @@ CURLcode Curl_vsetopt(struct Curl_easy *data, CURLoption option, va_list param)
     data->set.no_signal = (0 != va_arg(param, long)) ? TRUE : FALSE;
     break;
 
-  case CURLOPT_SPEEDTEST_MODE:
-    data->set.speedtest_mode = (0 != va_arg(param, long)) ? TRUE : FALSE;
+  case CURLOPT_DL_SPEEDTEST_MODE:
+    data->set.dl_speedtest_mode = (0 != va_arg(param, long)) ? TRUE : FALSE;
     break;
 
+  case CURLOPT_UL_SPEEDTEST_MODE:
+    data->set.ul_speedtest_mode = (0 != va_arg(param, long)) ? TRUE : FALSE;
+    break;
+
+
   case CURLOPT_SHARE:
   {
     struct Curl_share *set;
diff --git a/lib/transfer.c b/lib/transfer.c
index d7e231a01..246824152 100644
--- a/lib/transfer.c
+++ b/lib/transfer.c
@@ -602,8 +602,8 @@ static CURLcode readwrite_data(struct Curl_easy *data,
     }
 
     if(bytestoread) {
-      /* discard payload data in speedtest mode */
-      if (!k->header && data->set.speedtest_mode) {
+      /* discard payload data in dl speedtest mode */
+      if (!k->header && data->set.dl_speedtest_mode) {
         nread = recv(conn->sockfd, buf, bytestoread, MSG_TRUNC);
         result = (nread < 0) ? CURLE_READ_ERROR : CURLE_OK;
       } else {
diff --git a/lib/urldata.h b/lib/urldata.h
index c23437b51..10d63b2e9 100644
--- a/lib/urldata.h
+++ b/lib/urldata.h
@@ -1821,8 +1821,10 @@ struct UserDefined {
   BIT(reuse_forbid);   /* forbidden to be reused, close after use */
   BIT(reuse_fresh);    /* do not re-use an existing connection  */
   BIT(no_signal);      /* do not use any signal/alarm handler */
-  BIT(speedtest_mode); /* SpeedTest Mode, i.e discard read data and
-                          set performance flags */
+  BIT(dl_speedtest_mode); /* Downlink SpeedTest Mode, i.e discard read data and
+                             set performance flags */
+  BIT(ul_speedtest_mode); /* Uplink SpeedTest Mode, i.e use zerocopy and
+                             MSG_DONTWAIT */
   BIT(tcp_nodelay);    /* whether to enable TCP_NODELAY or not */
   BIT(ignorecl);       /* ignore content length */
   BIT(connect_only);   /* make connection, let application use the socket */
-- 
2.37.0.rc0.161.g10f37bed90-goog

